### DISCUSSION TITLE
Position Error when Drop a Node!

### DISCUSSION THREAD

**Discussion by Yeenyeong** [bcakmakoglu/vue-flow#239] [Q&A] [answered]
This is an example that is almost the same as the example on https://vueflow.dev/examples/dnd.html.
But when I drag a node to the position marked by the red rectangular box and drop the node, the node will go to the faraway position as the picture shows.
<img width="1364" alt="image" src="https://user-images.githubusercontent.com/50094666/180602300-7e9dd6ce-b1f4-43b9-8ec0-6d1ec8213562.png">
I think I've solve properly the relation of the node size and the mouse position. My code is following.

` 

      const nodeSize = {
        width: 320,
        height: 100
      };      

      console.log(event.clientX);

      console.log(event.clientY);

      const position = project({ 
        x: event.clientX - nodeSize.width, 
        y: event.clientY - nodeSize.height,
      });

      const newNode = {
        id: nodeID,
        type: 'custom',
        data: nodeData,
        position,
      };

      addNodes([newNode])`
How can I solve the problem?

**Comment by bcakmakoglu** [comment]
Your calculation is wrong.

The X and Y coordinates of your mouse minus the *whole* width and height of your node would not place it where you dropped it.
It would place it below and to the left of where you drop it (as you subtract the height of your node and the width).

Try dividing the node dimensions (width/height) by 2 to get the node centered at the mouse drop position.
```ts
  const position = project({ 
    x: event.clientX - (nodeSize.width / 2), 
    y: event.clientY - (nodeSize.height / 2),
  });

  const newNode = {
    id: nodeID,
    type: 'custom',
    data: nodeData,
    position,
	...nodeSize,
  };

**Reply by Yeenyeong** [reply]
Actually I've already tried dividing width and height by 2. It didn't work well.
And the position error seems so large(several times the size of the node) that I don't think it is caused by calculating the position improperly.

**Reply by bcakmakoglu** [reply]
Try this instead

```ts
  const flowbounds = wrapper.value.$el.getBoundingClientRect()

  const position = project({
    x: event.clientX - flowbounds.left,
    y: event.clientY - flowbounds.top,
  })

  const newNode = {
    id: getId(),
    type,
    position,
    label: `${type} node`,
    ...nodeSize,
  } as Node
  addNodes([newNode])
```

```vue
<template>
	<VueFlow v-model="elements" ref="wrapper" />
</template>

**Reply by Yeenyeong** [reply]
Can I achieve it in javascript? My project is not in typescript.

**Reply by bcakmakoglu** [reply]
It's the same in JS except that you just remove `as Node` 😄

**Reply by Yeenyeong** [reply]
But how can I get the wrapper variable?

**Reply by bcakmakoglu** [reply]
```js
const wrapper = ref()

...
```

```vue
<template>
	<VueFlow v-model="elements" ref="wrapper" />
</template>

**Reply by Yeenyeong** [reply]
Oh yeah! It works well ! Thanks for your patient help!

**Comment by Yeenyeong** [comment, marked as answer]
> Try this instead
> 
> ```ts
>   const flowbounds = wrapper.value.$el.getBoundingClientRect()
> 
>   const position = project({
>     x: event.clientX - flowbounds.left,
>     y: event.clientY - flowbounds.top,
>   })
> 
>   const newNode = {
>     id: getId(),
>     type,
>     position,
>     label: `${type} node`,
>     ...nodeSize,
>   } as Node
>   addNodes([newNode])
> ```
> 
> ```
> <template>
> 	<VueFlow v-model="elements" ref="wrapper" />
> </template>
> ```

Here is the answer by @bcakmakoglu.
Sincere Appreciation!
